pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/ReplicaGroupRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,573 @@
+          && removedServers.size() % targetNumInstancesPerPartition == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_REPLICA_GROUP;
+      }
+    } else if (sameNumReplicaGroup) {
+      if (oldNumInstancesPerPartition < targetNumInstancesPerPartition && isRemovedServersSizeZero
+          && !isAddedServersSizeZero && addedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.ADD_SERVER;
+      } else if (oldNumInstancesPerPartition > targetNumInstancesPerPartition && !isRemovedServersSizeZero
+          && isAddedServersSizeZero && removedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_SERVER;
+      }
+    }
+    return ReplicaGroupRebalanceType.UNSUPPORTED;
+  }
+
+  /**
+   * Rebalances segments for a replica group table without update
+   * @param idealState old idealstate
+   * @param tableConfig a table config
+   * @param replicaGroupPartitionAssignment a replica group partition assignment
+   * @param rebalanceGroupBootstrap bootstrap mode
+   * @return a rebalanced idealstate
+   */
+  private IdealState rebalanceSegments(IdealState idealState, TableConfig tableConfig,
+      ReplicaGroupPartitionAssignment replicaGroupPartitionAssignment, boolean rebalanceGroupBootstrap) {
+    Map<String, Map<String, String>> segmentToServerMapping = idealState.getRecord().getMapFields();
+    Map<String, LinkedList<String>> serverToSegments =
+        buildServerToSegmentMapping(segmentToServerMapping, rebalanceGroupBootstrap);
+
+    // Compute added and removed servers
+    List<String> oldServerInstances = new ArrayList<>(serverToSegments.keySet());
+    List<String> serverInstances = replicaGroupPartitionAssignment.getAllInstances();
+
+    List<String> addedServers = new ArrayList<>();
+    List<String> removedServers = new ArrayList<>();
+    for (String server : oldServerInstances) {
+      if (!serverInstances.contains(server)) {
+        removedServers.add(server);
+      }
+    }
+    for (String server : serverInstances) {
+      if (!oldServerInstances.contains(server)) {
+        addedServers.add(server);
+      }
+    }
+
+    // Add servers to the mapping
+    for (String server : addedServers) {
+      serverToSegments.put(server, new LinkedList<String>());
+    }
+
+    // Remove servers from the mapping
+    for (String server : removedServers) {
+      serverToSegments.remove(server);
+    }
+
+    // Check if rebalance can cause the downtime
+    Set<String> segmentsToCover = segmentToServerMapping.keySet();
+    Set<String> coveredSegments = new HashSet<>();
+    for (Map.Entry<String, LinkedList<String>> entry : serverToSegments.entrySet()) {
+      coveredSegments.addAll(entry.getValue());
+    }
+    if (!segmentsToCover.equals(coveredSegments)) {
+      LOGGER.warn("Segment is lost, the rebalance will cause downtime of the segment.");
```

Comment:
I don't understand why segments are lost. We should rebalance segments into other servers, right?

Reply:
In case of downsizing the replica group, we will need to re-distribute the segments from removed server. For example, if we use mirroring, taking away 3 servers with the same segment assignment will cause the inaccuracy of query result until the segments are redistributed because we won't have some segments available while rebalancing. I'm indicating that for the log. Do you think it's unnecessary?